### PR TITLE
Displays Consul version of each nodes in UI nodes section

### LINKED ISSUE #17097: Consul version of each nodes in UI nodes section
Hello guys,

It very useful to have the Consul version of each nodes in UI nodes section next to the IP address.
When doing a rolling upgrade or having federeted  setup, I feel like we need that.

Thx.

### PR DESCRIPTION
**Pull request by vijayraghav-io**
fixes #17097 Consul version of each nodes in UI nodes section

@jkirschner-hashicorp @huikang @team @Maintainers

Updated consul version in the request to register consul.
Added this as Node MetaData.
Fetching this new metadata in UI

Screenshot 2023-06-15 at 4 21 33 PM

Also made this backward compatible and tested.

Backward compatible in this context means - If consul binary with above PR changes is deployed to one of node, and if UI is run from this node, then the version of not only current (upgraded) node is displayed in UI , but also of older nodes given that they are consul servers only.
For older (non-server or client) nodes the version is not added in NodeMeta Data and hence the version will not be displayed for them.
If a old node is consul server, the version will be displayed. As the endpoint - "v1/internal/ui/nodes?dc=dc1" was already returning version in service meta. This is made use of in current UI changes.

Screenshot 2023-06-16 at 6 58 32 PM

### CODE REVIEW

**Comment by vijayraghav-io** [comment]
@maintainers @armon 
Reminding to review this PR

**Comment by jkirschner-hashicorp** [comment]
Hi @vijayraghav-io,

Thank you for suggesting a design and contributing an implementation! Upon considering your proposal, we realized that depending on the use case, there might be other related changes (or possibly alternative changes) worth considering.

Before we proceed with a technical review, we'd love to better understand the use case(s) that you and @Lord-Y both had that inspired your interest in #17097, hence our follow-up [here](https://github.com/hashicorp/consul/issues/17097#issuecomment-1595331823). For example, it seems like https://github.com/hashicorp/consul/issues/17146 might be better addressed by some kind of aggregate display of versions across agents in the future (which could co-exist with the display of this information on the node list view page).

Depending on your use case(s), it's also possible the node list page may or may not benefit from being able to use the search bar to filter on nodes running matching Consul versions (e.g., typing "1.15" filters to show only nodes running 1.15.x versions).

We want to make sure we consider this proposed design change in light of the underlying use case(s), that may or may not suggest other related changes to consider in the future.

Thanks again for engaging / collaborating with us!

**Comment by vijayraghav-io** [comment]
Hi @jkirschner-hashicorp 

Thanks for your reply! 
The idea to display consul version in node list page, is to have ability to filter nodes based on version as you mentioned. Also to have sorting ability as well "New to older" similar to "Healthy to UnHealthy" sort. This also provides a aggregated view.
Filtering and sorting can also be implemented in this PR.

**Comment by vijayraghav-io** [comment]
Hi @jkirschner-hashicorp @cizh 
Here is proposed design and implementation of version search with filter.
Screenshot 2023-06-22 at 1 02 09 PM

This provides a aggregated view w.r.t versions.
The filter dropdown lists current and two previous minor versions (dynamically), this as per [doc](https://developer.hashicorp.com/consul/docs/upgrading#large-version-jumps).
... lists all other older versions if any.

Note: Let me know, if this is ok, need to update some tests and optimize before review

**Comment by jkirschner-hashicorp** [comment]
@vijayraghav-io : Thanks for proposing a design and implementation allowing the search text to filter on version and adding a version dropdown! We're taking a look at the design. As a part of that, I have a follow-up question.

> ... (3 dots in menu ) -> lists all other/older versions if any.

**Can you help me understand what the dropdown would show in the following case?**
- The Consul agent serving the UI is version 1.15
- The datacenter also has Consul agents on the following versions:
  - 1.16
  - 1.14
  - 1.10
  - 1.7

_Note: We would not recommend operating a datacenter with this range of versions. I'm using it as a case to understand what the dropdown would show._

And out of curiosity, what was your thinking on using "..." to expand the list versus displaying all versions in use without needing an extra click? I suppose it depends on how many versions we reasonably expect someone might have in their datacenter. (It _shouldn't_ be more than 3 due to the general N-2 compatibility, though I've observed scenarios where there are a few more versions than that in use nonetheless.)

**Comment by vijayraghav-io** [comment]
@jkirschner-hashicorp 
For the above use case, the version dropdown lists as below:
   1.16.x
   1.15.x
   1.14.x
   ...

Explanation:
1. Three latest versions in the cluster/datacenter are listed and then ...
2. When user clicks on ... , the nodes with version 1.10 and 1.7 are only filtered (considering given use case)
3. The idea of having ... (or 3dots) in the menu had two reasons
- Limit the menu list items (say to max 4). 
- With earlier approach of using NodeMeta to have version data , there are scenarios where version info cannot not be retrieved for some of the agents in cluster unless they are running as servers and the agent serving UI is running as Leader. ... -> served the purpose of filtering such nodes where version info could not be retrieved. But now, the approach is to use "/v1/agent/members" endpoint to retrieve all the serf members in the cluster and get version from its "build" tag. With this approach there will be No cases of version info not retrieved for any agent.
- So ... can be discarded if its better idea to show all available distinct versions (upto minor ex:- 1.17.x , 1.16.x ,..,1.10.x,..) in the cluster.

File: agent/consul/state/catalog.go
Comment (huikang) [inline agent/consul/state/catalog.go:3456]: can we consider including the change in our unit test case `TestStateStore_NodeInfo_NodeDump` in `catalog_test.go`?

File: agent/consul/leader.go
Comment (huikang) [inline agent/consul/leader.go:1111]: should `buildVersion.String()` work as well?

**Review by huikang — COMMENTED** [review]
I only reviewer the changes under `./agent`, so still need some UI eng to look at the PR. Thanks, @vijayraghav-io

**Comment by vijayraghav-io** [comment]
@huikang please hold on! updating some changes, will confirm when ready for review. Thank you!

**Comment by jkirschner-hashicorp** [comment]
@cizh and I will follow up soon on which versions / wording to use in the dropdown, as well as the mockup of a sort by latest/oldest. I'd recommend waiting for us to follow up before making changes (so you can batch any changes at once).

**Comment by cizh** [comment]
image
Hi @vijayraghav-io Please see above for a mockup with "filter" function: after discussion with @jkirschner-hashicorp, we decided displaying **all** of the user's current versions in the filter options would make the logic more clear to users. Although it'll be great UX if we can digest for our users, at this time, the number is the least confusing because some versions may be compatible with much earlier versions but not the latest so we can't precisely say they are out of support.

Here is a mockup for "sort" function that we would recommend if you have the time and the want to build it.
image

**Comment by vijayraghav-io** [comment]
Hi @cizh @jkirschner-hashicorp @huikang - PR is ready for review.
Consul version are displayed for all nodes, both sorts and filters for versions are implemented.

Screenshot 2023-06-29 at 10 23 12 PM
Screenshot 2023-06-29 at 10 23 12 PM

File: agent/consul/leader.go
Comment (vijayraghav-io) [reply agent/consul/leader.go]: Yes updated

**Comment by vijayraghav-io** [comment]
Hi @huikang @jkirschner-hashicorp , please find below the test reports:
Screenshot 2023-06-30 at 9 53 55 AM
Screenshot 2023-06-30 at 8 24 39 AM
Screenshot 2023-06-30 at 8 21 17 AM

File: agent/consul/state/catalog.go
Comment (vijayraghav-io) [reply agent/consul/state/catalog.go]: @huikang , Please note - this change is not there now. The above approach would work only if agent is running as server & Leader (due to RPC Forward).
So, now a better robust approach is taken to get consul version irrespective of agent running as client or server and non-leader/leader. The approach is to get all serf members and get version from 'build' tag. Hence above specific code change is discarded and therefore the testcase.

Comment (vijayraghav-io) [reply agent/consul/state/catalog.go:3459]: @huikang , this was added as fallback, if for some reason not able to get version from serf members.
Updated the test case in TestStateStore_NodeInfo_NodeDump.
Screenshot 2023-07-01 at 7 05 13 PM

**Comment by vijayraghav-io** [comment]
@jkirschner-hashicorp @huikang @cizh - Reminding to please review this PR, have attached the test snapshots wherever required.

**Review by cizh — APPROVED** [review]
The new user interfaces are aligned with the design and solves the user needs. Thank you for making this change.

File: .changelog/17754.txt
Comment (david-yu) [inline .changelog/17754.txt:2]:
```suggestion
ui: Display the Consul agent version in the nodes list, and allow filtering and sorting of nodes based on versions.
```

**Comment by david-yu** [comment]
@vijayraghav-io It looks like this change introduces anti-entropy sync errors based on the 5 tests or so that are failing. I'm not sure how to resolve them but we need to make sure the tests are passing before we can merge.

File: .changelog/17754.txt
Comment (vijayraghav-io) [reply .changelog/17754.txt]: @david-yu , updated

**Comment by vijayraghav-io** [comment]
Thank you! @cizh @david-yu. 

@david-yu there were some dependent test cases in other modules, have updated these and tested, verified. 
Din't run complete tests earlier, had only run relevant tests.
Now ran a complete test locally after update, its fine!.

**Review by WenInCode — APPROVED** [review]
LGTM, one suggestion but overall a welcome improvement. Maybe worth getting another review from someone on the backend core team. I'll send a ping.

File: ui/packages/consul-ui/app/components/consul/node/search-bar/index.hbs
Comment (vijayraghav-io) [reply ui/packages/consul-ui/app/components/consul/node/search-bar/index.hbs:120]: @WenInCode , sure enabled multi-selection for version filter

**Review by huikang — APPROVED** [review]

**Comment by jkirschner-hashicorp** [comment]
Hi @vijayraghav-io,

To give you some context on the reviews you just received: all necessary stakeholders have now approved your PR!
- Design (@cizh)
- Front-end engineering (@WenInCode)
- Back-end engineering (@huikang)

It looks like @huikang's approval came with a nit-suggestion to slightly clean up a comment. After that small clean-up, a maintainer can proceed with the merge.

This looks great - thanks again for the collaboration and contribution!

File: agent/consul/state/catalog.go
Comment (jkirschner-hashicorp) [reply agent/consul/state/catalog.go:3453]: @huikang , @vijayraghav-io : What about changing it this way? (Made it as a Github suggestion so the commit is easy to make)

```suggestion
			// If version doesn't exist in node meta, set it from the Consul service meta
```

Comment (vijayraghav-io) [reply agent/consul/state/catalog.go]: @huikang @jkirschner-hashicorp , committed the suggested changes.

**Comment by vijayraghav-io** [comment]
Thank you! @jkirschner-hashicorp , @huikang

**Comment by vijayraghav-io** [comment]
Committed the suggestion. 
Thank you! @jkirschner-hashicorp , @huikang.

**Comment by johnnyjeannatasha** [comment]
> @jkirschner-hashicorp @huikang @cizh - Reminding to please review this PR, have attached the test snapshots wherever required.

Bye bye.

**Pull request merged**
